Source/Core/Core/IOS/DolphinDevice.cpp

```diff
@@ -45,9 +45,11 @@ IPCReply GetSystemTime(const IOCtlVRequest& request)
     return IPCReply(IPC_EINVAL);
   }
 
-  const u32 milliseconds = Common::Timer::GetTimeMs();
+  auto milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+                          std::chrono::system_clock::now().time_since_epoch())
+                          .count();
 
-  Memory::Write_U32(milliseconds, request.io_vectors[0].address);
+  Memory::Write_U32(static_cast<u32>(milliseconds), request.io_vectors[0].address);
```

Comment:
2 things here:
* can we freely change the ioctl to return 8bytes instead of 4?
* what is _really_ wanted to be returned here? It's named "system time", but was previously returning QPC output on windows, something else on linux and macos. Just a monotonic counter? Just `system_clock`? The **real** output of `GetLocalTimeSinceJan1970`? TZ-adjust `system_clock`?

Reply:
After reading the original PR https://github.com/dolphin-emu/dolphin/pull/8445 i moved back to using `Common::Timer::NowMs` here, because that ensures it's the same clock that dolphin is using to throttle.

Reply:
> Or maybe start a timer when /dev/dolphin is created and only return ElapsedMs in the ioctl?

I decided to implement this. I'd say PR is ready now.

Reply:
Changing that return would break backwards compatibility, so it should be avoided.

I did a check for projects on GitHub that use `/dev/dolphin`; here's the results:
* [SeekyCt/spm-practice-codes](https://github.com/SeekyCt/spm-practice-codes): only uses it to [check if Dolphin is present](https://github.com/SeekyCt/spm-practice-codes/blob/f524a330c5291226a8bc8016d72ac78c57d124ba/source/mod.cpp#L124-L143), and if so [patches exiting the sound system](https://github.com/SeekyCt/spm-practice-codes/blob/f524a330c5291226a8bc8016d72ac78c57d124ba/source/mod.cpp#L168-L172) (to work around [issue 12932](https://bugs.dolphin-emu.org/issues/12932), [changes the font scale on the crash screen](https://github.com/SeekyCt/spm-practice-codes/blob/82337fc59b21eb42feb44c89e555fffdc8a80833/source/exception.cpp#L131) (under the assumption that Dolphin is always using the free font), and [allows patching some text](https://github.com/SeekyCt/spm-practice-codes/blob/f524a330c5291226a8bc8016d72ac78c57d124ba/source/miscmenu.cpp#L48-L53) ([other context](https://github.com/SeekyCt/spm-practice-codes/blob/f524a330c5291226a8bc8016d72ac78c57d124ba/source/miscmenu.cpp#L48-L53)).
* [RiiConnect24/Mail-Patcher](https://github.com/RiiConnect24/Mail-Patcher): only uses it to [check if Dolphin is present](https://github.com/RiiConnect24/Mail-Patcher/blob/c0bf0a89842342881a20e9cde55ea56531c420a8/source/nand.c#L6-L13), and [if so reports an error message](https://github.com/RiiConnect24/Mail-Patcher/blob/c0bf0a89842342881a20e9cde55ea56531c420a8/source/main.c#L48-L56) (I don't think Dolphin emulates the mail functionality at all, so this makes sense).
* [WiiLink24/libpatcher](https://github.com/WiiLink24/libpatcher): only uses it to [check if Dolphin is present](https://github.com/WiiLink24/libpatcher/blob/3e51f23ac170df9f3961306a10ac5007d91d4084/source/libpatcher.c#L18-L39), and if so [disables patching IOS and escalating permissions](https://github.com/WiiLink24/libpatcher/blob/3e51f23ac170df9f3961306a10ac5007d91d4084/source/libpatcher.c#L64-L79) (as Dolphin doesn't do IOS LLE). (I guess it's also worth noting that if we ever _did_ do full IOS LLE (including the IOS kernel), `/dev/dolphin` wouldn't exist unless we wrote a new IOS module for it, so this is reasonable.)
* [larsenv/GlowWii](https://github.com/larsenv/GlowWii): only uses it to [check if Dolphin is present](https://github.com/larsenv/GlowWii/blob/4a6ca98ebbe3a37cb968e7b2a6535efb225757ca/source/main.c#L49-L57), as [Dolphin doesn't emulate the disc drive light](https://github.com/dolphin-emu/dolphin/blob/dd67d5fbc4aeffe78daa7b1697ebf39bdd7ccd3e/Source/Core/Core/HW/WII_IPC.cpp#L218).
* [DacoTaco/priiloader](https://github.com/DacoTaco/priiloader): only uses it to [check if Dolphin is present](https://github.com/DacoTaco/priiloader/blob/5e75587798a80116a17f05ba4f64a4cab609a4c1/Installer/source/main.cpp#L1502-L1505) to [skip escalating permissions](https://github.com/DacoTaco/priiloader/blob/5e75587798a80116a17f05ba4f64a4cab609a4c1/Installer/source/main.cpp#L1573-L1576)
* [TheLordScruffy/mkw-tournament-museum](https://github.com/TheLordScruffy/mkw-tournament-museum): only uses it to [check if Dolphin is present](https://github.com/TheLordScruffy/mkw-tournament-museum/blob/9e1b1dfa900c5cc1699f222d3f782e66473e2a34/loader/loader.cpp#L221-L238). This function is never called.
* [leo60228/change-disc](https://github.com/leo60228/change-disc): [uses currently-nonexistent ioctls 0x06 and 0x07 to change discs](https://github.com/leo60228/change-disc/blob/a670dfe2d3ef65481fa86f0dee0ca64002348b7c/source/template.c#L76-L111); only intended to run in Dolphin; see PR #8636.
* [stblr/mkw-sp](https://github.com/stblr/mkw-sp): Probably the main interesting use of it. In the stub, it [uses ioctl 0x04](https://github.com/stblr/mkw-sp/blob/eca1a6c88dcfd3aadae40021ca09150586e6f263/stub/Dolphin.cc) to [set the speed limit](https://github.com/stblr/mkw-sp/blob/a2d2dc4803fc77a6952937fd852929ba4fb8ab73/stub/Stub.cc#L77-L82). In the main code, it [uses it to detect Dolphin](https://github.com/stblr/mkw-sp/blob/0505b427ace12cd8258776930e4f2576d38a7b1f/common/IOS.cc#L85-L99) to [skip permission escalation](https://github.com/stblr/mkw-sp/blob/0505b427ace12cd8258776930e4f2576d38a7b1f/common/IOS.cc#L137-L140), and [a full wrapper around all ioctls](https://github.com/stblr/mkw-sp/blob/eca1a6c88dcfd3aadae40021ca09150586e6f263/payload/sp/IOSDolphin.c), which it uses [for a self-test](https://github.com/stblr/mkw-sp/blob/19f411b7e8d260dae68f3868c4bdd92b1f12eaf0/payload/egg/core/eggSystem.c#L35-L79), to [check the version](https://github.com/stblr/mkw-sp/blob/a2d2dc4803fc77a6952937fd852929ba4fb8ab73/payload/sp/Host.c#L42-L69), and to adjust the speed limit ([here](https://github.com/stblr/mkw-sp/blob/8ef7f6449aedde404ab52336f62f7fdca680de69/payload/sp/Payload.cc#L134-L138), [here](https://github.com/stblr/mkw-sp/blob/67384c9036e912d5a01e830853da53adc7a7bf06/payload/game/host_system/SystemManager.cc#L11-L57), and [here](https://github.com/stblr/mkw-sp/blob/185586b081de4868658e0235b47c70df20107d79/payload/egg/core/eggSceneManager.cc#L15-L101)). Outside of the test it doesn't seem to use `IOSDolphin_GetSystemTime` for anything.

Based on this, you could maybe change it without breaking any open-source homebrew, but it's probably better to avoid doing so. You could theoretically allow the output size to be either 4 or 8 and choose to write a u32 or u64 based on that, which would be backwards-compatible, but a bit awkward.

@Leseratte10: Any thoughts on this?

Reply:
The main reason I added that Ioctl was for Wiimmfi, Mario Kart Wii (when running on Wiimmfi) checks that value to figure out if Dolphin is lagging and to make sure a bunch of ingame timers are running at the correct speed even if Dolphin lags. Just changing that thing to return a 64-bit value might break that so it'd need a Wiimmfi update. 

But given that the code currently returns an error if the output size is anything but 4, I would also just suggest leaving the current behaviour (write 32-bit) if the output size is 4, and write a 64-bit value if the output size is 8. 

I'm not sure what currently happens in the Wiimmfi payload when that value rolls over and if Wiimmfi's code would break if this thing were to change to return 64-bit instead. I'll try to figure that out and let you know.

Reply:
For what it's worth, if the values are used like this:

```C++
u32 start = IOSDolphin_GetSystemTime();
// ... do some work, or wait until the next frame
u32 end = IOSDolphin_GetSystemTime();
u32 duration = end - start;
```

this will behave correctly even if an overflow occurs (as unsigned overflow is defined behavior, though signed overflow is undefined behavior (in practice I think you'd get the same result, though?)). It's similar to using [`gettick()` instead of `gettime()`](https://github.com/devkitPro/libogc/blob/bc4b778d558915aa40676e33514c4c9ba2af66b8/libogc/timesupp.c#L26-L54); `gettick()` can wrap around and you'll still be fine as long as you're subtracting. You would have trouble if you checked `start <= end` though.

Reply:
Just checked the Wiimmfi source code, there's two `int`s, I call IOS_Ioctlv once at race start, and then every in-game frame, and use this to figure out the actual real time amount of frames that should have passed: 
`int real_clock_framecount = ((current_time_real - race_time_start_realtime) / (float)16.6666666667);`

So that would be a signed int, but I can easily make that an unsigned int instead if that's less likely to cause issues. 

I checked the generated assembly - it's loading both values into a register then just uses `subf`, so that should be fine regarding overflow.

Reply:
In theory a signed int could cause problems if `race_time_start_realtime` was, say, `2147483392` (`0x7fffff00`) and `current_time_real` was `2147483904` (`0x80000100`, which would be `-2147483392` signed), because `-2147483392 - 2147483392 = -4294966784` which is an overflow (but note that `-4294966784 & 0xffffffff` is `512`). In practice, the compiler isn't going to care since actually checking for that overflow would be extra work (but [apparently it can optimize under the assumption that it won't happen](https://stackoverflow.com/a/16188924), or do anything it wants in that case). Using an unsigned int [guarantees the wrapping](https://stackoverflow.com/a/16188846) (and it also doesn't make sense for the frame count to be negative in the first place).

As for 32-bit versus 64-bit, `0x7fffffff` milliseconds is about 25 days (after which you'll always get an overflow when subtracting), and `0xffffffff` milliseconds is about 50 days. A race won't go anywhere near that long (unless you decide to do 500,000 laps of Baby Park). So I don't think 64-bit times are needed for any normal use-case (and remember that the Wii is a 32-bit console, so 64-bit values have a larger overhead).

Reply:
> In theory a signed int could cause problems ...

Yes, but that is an issue with Wiimmfi unrelated to this PR. This PR should help if anything, since the returned value starts near zero instead of some "random" value.

Reply:
I tweaked the comment in DolphinDevice a little, otherwise haven't changed behavior.